### ISSUE TITLE
Hangs with Python 3.7

### ISSUE THREAD

**Issue by cjltsod** [pahaz/sshtunnel#138]
I'm using SSHTunnelForwarder with Python 3.7. After finishing my work, I run `server.stop()` and hangs.

I believe it caused by `socketserver.ThreadingMixIn`
https://docs.python.org/3.7/library/socketserver.html#socketserver.ThreadingMixIn
> Changed in version 3.7: socketserver.ForkingMixIn.server_close() and socketserver.ThreadingMixIn.server_close() now waits until all child processes and non-daemonic threads complete. Add a new socketserver.ForkingMixIn.block_on_close class attribute to opt-in for the pre-3.7 behaviour.

Here's my work around. I changed the attribute `server._server_list[0].block_on_close = False`, and it behave like usual. It closed normally and not being hang.

 I'm not sure if setting `block_on_close = False` is a good way or not.

Issue #135 may have the same cause.

**Comment by cjltsod** [comment]
Here's what I work around. https://github.com/cjltsod/sshtunnel/commit/2e56a5b839c627472d3feae98196808ff9aca43e
Still not sure if it's a good way or not.

**Comment by fernandezcuesta** [comment]
Hi, thanks for the valuable information.
If I understand properly, when `daemon_threads` is `True` and `server_close()` is executed it will not wait for the threads created by `ThreadingMixIn` to be closed, which is the same as `block_on_close = False`.
My guess is that this parameter is more focused on unifying both `ForkingMixIn` and `ThreadingMixIn` behaviour rather than changing how things worked until now.
Did you try setting `DAEMON = True` or just `daemon_threads = True`? Do you obtain the same results?

**Comment by fernandezcuesta** [comment]
Well maybe it's much easier just setting `tunnel.daemon_forward_servers = True` before starting the tunnel rather than patching the constant.

**Comment by alejoar** [comment]
So are we getting a release with this fix or not? sshtunnel currently doesn't work properly in Python 3.7+

**Comment by rkkautsar** [comment]
Still hangs in 3.7+.
I tried setting `daemon_forward_servers = True` before starting the forwarder but it still doesn't work..

**Comment by cjltsod** [comment]
It has been patched with v0.1.5 at https://github.com/pahaz/sshtunnel/pull/149/commits/b2e0cc310dcde37be13f239a7b4379d46f34dd34 which is not on pypi yet.
It has been merged into master branch, so you can simply install through GitHub by using 
`pip install git+https://github.com/pahaz/sshtunnel.git`

**Comment by fernandezcuesta** [comment]
updated [PYPI](https://pypi.org/project/sshtunnel/)

**Comment by rkkautsar** [comment]
Any plan to also update the conda-forge package?

**Comment by eannam** [comment]
Hi, I am still getting this issue with python 3.7 using v0.1.5.

**Comment by descello** [comment]
still hang issue with python 3.7.5 using v0.1.5.

**Comment by pyMan** [comment]
Hi, I am still getting this issue with python 3.8 using v0.1.5.

**Comment by PozziSan** [comment]
> Hi, I am still getting this issue with python 3.8 using v0.1.5.
> 
> EDIT: specifically it happens to me only if I execute any kind of query through the ssh tunnel (I'm using django 2.2.9)

Hi! I'm getting the same here, after executing queries with pandas, on a flask application using sqlAlchemy.
The application is running with Python 3.8 usign v0.1.5.

**Comment by dem82** [comment]
I've solved the hanging problem with #72 
Python 3.7 using v0.1.5

**Comment by erfannariman** [comment]
Still getting this issue with `python 3.7.7`, after using `pandas.to_sql` and creating an engine with `sqlalchemy`. Using `engine.dispose` and `server.daemon_forward_servers = True` does not work.

**Comment by audciz** [comment]
Try adding `set_keepalive=120` to sshtunnel initialization parameters. Seems like shhtunnel is already closed when next action is performed. Solved my problem with psycopg2

**Comment by youngsoul** [comment]
Still hangs when using as context manager:

Env:
Python 3.7.3
SSHTunnelForwarder 0.1.5
OS: macOS Catalina

Gist example: https://gist.github.com/youngsoul/a02e45ef48e405c148f63bcfaea5ad8f

I have found that SSHTunnelForwarder will hang if I use it as a context manager using 'with'.  In the gist above see the function, 'run_query_with'.  If I create the SSHTunnelForwarder and manage the start/stop myself then it does not hang.  In the gist see the function, 'run_query'.

I did not have to set the flags as suggested above in 'run_query' and it is a little bit of mystery as to why that would be.

As far as I can tell, you definitely cannot set those properties if using as a context manager, because the __enter__ setups up the transport and the flags are already set with _DAEMON.

**Comment by tdstarkisoc** [comment]
> Still getting this issue with `python 3.7.7`, after using `pandas.to_sql` and creating an engine with `sqlalchemy`. Using `engine.dispose` and `server.daemon_forward_servers = True` does not work.

@erfannariman did you ever find a solution to this?

**Comment by youngsoul** [comment]
@tdstarkisoc While I have tried very recently, the gist I wrote shows a work around that does not use a context manager.

Gist example: https://gist.github.com/youngsoul/a02e45ef48e405c148f63bcfaea5ad8f

Running without the context manager, as shown in 'run_query' method worked for me. With the context manager, as shown in 'run_query_with' method did NOT work for me.

**Comment by tdstarkisoc** [comment]
After much trial and error, I found some workarounds.

Generally, I seemed to get some success by leveraging the `poolclass=StaticPool` feature in the SQLAlchemy engine. I am not using SSHTunnelForwarder as a context manager. That didn't seem to work. I have it defined as a variable. I am also using `server.daemon_forward_servers = True` and `server.daemon_transport = True`. 

When I finally converted the server to a class-based context manager, the code was hanging again. I then tried using `method=multi` as part of my Pandas to_sql statements and that seemed to make things work again.

**Comment by pahaz** [comment]
1) `sshtunnel` is always waiting until closing all connections during the `stop()` called.
2) If you don't want to wait, you can use `.stop(force=True)`
3) You can have a problem with hangs if you don't call `.stop`. For example, you have some unhandled exception in `MainThread` and you don't call `.stop` in this case. As a workaround you can use this hack:
```
# I don't recommend to use this hack!
# Hack threadings https://docs.python.org/3/library/threading.html#threading.Thread.daemon ...
sshtunnel.SSHTunnelForwarder.daemon_forward_servers = True
sshtunnel.SSHTunnelForwarder.daemon_transport = True
```

Or, It's **better** to use `with open_tunnel(...)` which always stops the tunnel.

Do anyone have an issue related to hangs with `sshtunnel==0.3.1` and `.stop(force=True)` ?

**Comment by tdstarkisoc** [comment]
FWIW, I was trying to use open_tunnel and I was still having issues with `.stop(force=True)`.

**Comment by erfannariman** [comment]
> > Still getting this issue with `python 3.7.7`, after using `pandas.to_sql` and creating an engine with `sqlalchemy`. Using `engine.dispose` and `server.daemon_forward_servers = True` does not work.
> 
> @erfannariman did you ever find a solution to this?

I did and I searched through my projects to find it, but I can't find which project it was. If I bump into, will definitely post it here.

**Comment by pahaz** [comment]
@tdstarkisoc could you please provide a reproducible example?

**Comment by tdstarkisoc** [comment]
@pahaz Unfortunately, I don't have that version anymore, but I wasn't able to find a solution to the hanging in any instance where I used it as a context manager. I experimented with `.stop(force=True)` in each instance.

**Comment by tdstarkisoc** [comment]
This is is what I have currently as a working version:

`class DatabaseConn:
    def __init__(self, database):
        self.database = database
        self.engine = ""
        self.server = SSHTunnelForwarder(##.##.##.##,
                                         ssh_port=22,
                                         ssh_username=#####,
                                         ssh_password=#####,
                                         remote_bind_address=('localhost', 7777))

    def __enter__(self):
        logging.info("Attempting to establish SSH connection.")

        # Server connection config
        self.server.daemon_forward_servers = True
        self.server.daemon_transport = True
        self.server.start()

        # Database connection variables
        user = #####
        password = #####
        local_port = str(self.server.local_bind_port)
        conn_string = f"databasetype://{user}:{password}@127.0.0.1:{local_port}/{self.database}"

        # Database engine
        self.engine = create_engine(conn_string,
                                                     poolclass=StaticPool,
                                                     echo=True,
                                                     echo_pool=True)
        logging.info("SSH connection established.")
        return self.engine

    def __exit__(self, exc_type, exc, exc_tb):
        logging.info("Attempting to close SSH connection.")
        self.engine.dispose()
        self.server.stop(force=True)
        logging.info("SSH connection closed.")`

**Comment by pahaz** [comment]
@tdstarkisoc are you sure that

```
        self.server.daemon_forward_servers = True
        self.server.daemon_transport = True
```

hack is required for you? did you test it on `sshtunnel==0.3.1`?

**Comment by tdstarkisoc** [comment]
> @tdstarkisoc are you sure that
> 
> ```
>         self.server.daemon_forward_servers = True
>         self.server.daemon_transport = True
> ```
> 
> hack is required for you? did you test it on `sshtunnel==0.3.1`?

@pahaz I commented those lines and didn't appear to have an issue on Windows 10, but I did have an issue on Ubuntu 20.04 LTS with `sshtunnel==0.2.2`. The update seemed to fix the hanging issue related to closing the connections and I can now use single insert statements with to_sql. Thank you!

**Comment by pahaz** [comment]
@tdstarkisoc thanks you for feedback and tests.

I'll close the issue. And if anyone have an issue related to hangs with `sshtunnel==0.3.1` and `.stop(force=True)` please **open a new issue**. And read this first: https://github.com/pahaz/sshtunnel/issues/138#issuecomment-728200814

And also we have a plan to change the default threading daemon option for sshtunnel threads. 
But we needed more tests for those cases.
